src/libopenrave/kinbody.cpp

```diff
@@ -5151,6 +5151,37 @@ bool KinBody::IsAttached(const KinBody &body) const
     return _IsAttached(body.GetEnvironmentBodyIndex(), vAttachedVisited);
 }
 
+void KinBody::GetDirectlyAttached(std::vector<KinBodyPtr>& vAttached) const
+{
+    // Clear output and prepare
+    vAttached.clear();
+    vAttached.reserve(_listAttachedBodies.size());
+
+    // The list of attached bodies is not unique, so deduplicate as we go
+    std::unordered_set<const KinBody*> visitedBodies;
+
+    // For each weakly attached body, check if it is still present and if so add it to the output if we haven't already
+    for (const KinBodyWeakPtr& weakAttachedBody : _listAttachedBodies) {
+        // Ignore any bodies that are no longer live
+        KinBodyPtr attachedBody = weakAttachedBody.lock();
+        if (!attachedBody) {
+            continue;
+        }
+
+        // Ignore any bodies that are already present in our visited set
+        const KinBody* const kinBodyRawPtr = attachedBody.get();
+        if (visitedBodies.find(kinBodyRawPtr) != visitedBodies.end()) {
```

Comment:
@rschlaikjer isn't `_listAttachedBodies` supposed to be guaranteed to be unique bodies? Why do you check for duplicates?

Also, why not create:
```
void KinBody::GetDirectlyAttachedEnvironmentBodyIndices(std::vector<int>& vAttached) const
```

so that body indices can be returned instead of pointers? Should be faster, and the environment itself can get the pointer immediately by indexing `_vecbodies`

Reply:
The comment on that field claims that it _can_ have duplicates; is that incorrect?
```
d071d0458a (rdiankov                2009-12-23 03:40:10 +0000)|  3685     std::list<KinBodyWeakPtr> _listAttachedBodies; ///< list of bodies that are directly attached to this body (can have duplicates)                                      ```

Re: methods, I forgot the env can just access `_listAttachedBodies`, so I have changed the code to iterate that directly rather than scanning / creating a new vector.